packages/utils/telemetry-utils/src/config.ts

```diff
@@ -0,0 +1,263 @@
+ */
+export class CachedConfigProvider implements IConfigProvider {
+    private readonly configCache = new Map<string, StronglyTypedValue>();
+    private readonly orderedBaseProviders: (IConfigProviderBase | undefined)[];
+
+    constructor(
+        ... orderedBaseProviders: (IConfigProviderBase | undefined)[]
+    ) {
+        this.orderedBaseProviders = [];
+        const knownProviders = new Set<IConfigProviderBase>();
+        const candidateProviders = [...orderedBaseProviders];
+        while (candidateProviders.length > 0) {
+            const baseProvider = candidateProviders.shift()!;
+            if (baseProvider !== undefined
+                && isConfigProviderBase(baseProvider)
+                && !knownProviders.has(baseProvider)
+            ) {
+                knownProviders.add(baseProvider);
+                if (baseProvider instanceof CachedConfigProvider) {
+                    candidateProviders.push(...baseProvider.orderedBaseProviders);
+                } else {
+                    this.orderedBaseProviders.push(baseProvider);
+                }
+            }
+        }
+    }
+    getBoolean(name: string): boolean | undefined {
+        return this.getCacheEntry(name)?.boolean;
+    }
+    getNumber(name: string): number | undefined {
+        return this.getCacheEntry(name)?.number;
+    }
+    getString(name: string): string | undefined {
+        return this.getCacheEntry(name)?.string;
+    }
+    getBooleanArray(name: string): boolean[] | undefined {
+        return this.getCacheEntry(name)?.["boolean[]"];
+    }
+    getNumberArray(name: string): number[] | undefined {
+        return this.getCacheEntry(name)?.["number[]"];
+    }
+    getStringArray(name: string): string[] | undefined {
+        return this.getCacheEntry(name)?.["string[]"];
+    }
+
+    getRawConfig(name: string): ConfigTypes {
+        return this.getCacheEntry(name)?.raw;
+    }
+
+    private getCacheEntry(name: string): StronglyTypedValue | undefined {
+        if (!this.configCache.has(name)) {
+            for (const provider of this.orderedBaseProviders) {
+                const parsed = stronglyTypedParse(provider?.getRawConfig(name));
+                if (parsed !== undefined) {
+                    this.configCache.set(name, parsed);
+                    return parsed;
+                }
+            }
+            // configs are immutable, if the first lookup returned no results, all lookups should
+            this.configCache.set(name, { raw: undefined });
+        }
+        return this.configCache.get(name);
+    }
+}
+
+/**
+ * A type containing both a telemetry logger and a configuration provider
+ */
+export interface MonitoringContext<T extends ITelemetryBaseLogger = ITelemetryLogger> {
+    logger: T;
+    config: IConfigProvider;
+}
+
+export function loggerIsMonitoringContext<T extends ITelemetryBaseLogger = ITelemetryLogger>(
+    obj: T): obj is T & MonitoringContext<T> {
```

Comment:
You don't need to say `T &`, it knows it's still `T`.

```suggestion
    obj: T,
): obj is MonitoringContext<T> {
```